<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Analyze RNA-seq dataset by trans-PCO pipeline</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Trans-PCO</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="sim_null.html">Null simulations</a>
    </li>
    <li>
      <a href="sim_alt.html">Power comparison in simulations</a>
    </li>
    <li>
      <a href="rna_seq.html">Analyze a RNA-seq dataset</a>
    </li>
    <li>
      <a href="vis_distri_sig.html">Visualize trans signals</a>
    </li>
    <li>
      <a href="signal_cis.html">Cis genes of trans signals</a>
    </li>
    <li>
      <a href="coloc_gwas.html">Colocalization analysis of trans loci and GWAS</a>
    </li>
    <li>
      <a href="coloc_vis_prop.html">Colocalization proportion of trans-eQTLs</a>
    </li>
    <li>
      <a href="sldsc.html">Perform S-LDSC</a>
    </li>
    <li>
      <a href="vis_sldsc.html">Visualize the trait heritability enrichment</a>
    </li>
    <li>
      <a href="coloc_cis.html">Colocalization analysis of trans loci and cis genes</a>
    </li>
  </ul>
</li>
<li>
  <a href="ref.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liliw-w/Trans">
    <span class="fa fa-github"></span>
     
    Source
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Analyze RNA-seq dataset by trans-PCO pipeline</h1>

</div>


<div id="goal" class="section level2">
<h2>Goal</h2>
<p>To search for trans-eQTLs associated with gene co-expression modules in a real RNA-seq dataset <span class="citation">(Battle et al. 2014)</span> using trans-PCO pipeline.</p>
</div>
<div id="main-steps" class="section level2">
<h2>Main steps</h2>
<p>The workflow is written in snakemake (see bottom code). There are a few main steps, including,</p>
<ol style="list-style-type: decimal">
<li><p>Regress covariates out of gene expression profiles.</p></li>
<li><p>Cluster genes into gene modules with highly co-expressed genes.</p></li>
<li><p>Calculate z-scores of SNPs across the genome for all genes by tensorQTL.</p></li>
<li><p>Calculate p-values of each pair of (a SNP, a gene module) across all SNPs and gene modules using PCO test.</p></li>
<li><p>Permute sample labels to obtain the empirical null p-value distribution, in order to control FDR (&lt;10%) among p-values calculated in last step.</p></li>
<li><p>Identify significant (trans-eQTLs, gene module) pairs.</p></li>
</ol>
</div>
<div id="snakemake-workflow" class="section level2">
<h2>Snakemake workflow</h2>
<ul>
<li>Steps 1-2,</li>
</ul>
<pre class="bash"><code>configfile: &quot;config.yaml&quot;

rule all:
  input: &#39;result/&#39;+config[&#39;file_coexp_module&#39;]


rule covariates:
  input:
    file_gene_annotation=config[&#39;dir_mappability&#39;]+config[&#39;file_gene_annotation&#39;],
    file_mappability=config[&#39;dir_mappability&#39;]+config[&#39;file_mappability&#39;],
    file_cross_mappability=config[&#39;dir_mappability&#39;]+config[&#39;file_cross_mappability&#39;],
    file_covariates=config[&#39;dir_expression&#39;]+config[&#39;file_covariates&#39;],
    file_ex=config[&#39;dir_expression&#39;]+config[&#39;file_ex&#39;]
  output:
    file_gene_meta=&#39;result/&#39;+config[&#39;file_gene_meta&#39;],
    file_ex_var_regressed=&#39;result/&#39;+config[&#39;file_ex_var_regressed&#39;]
  script:
    &#39;script/&#39;+config[&#39;script_covariates&#39;]


rule coexp_module:
  input:
    file_ex_var_regressed=&#39;result/&#39;+config[&#39;file_ex_var_regressed&#39;]
  output:
    file_coexp_module=&#39;result/&#39;+config[&#39;file_coexp_module&#39;],
    file_Nmodule=&#39;result/Nmodule.txt&#39;
  params: minModuleSize=config[&#39;minModuleSize&#39;]
  script:
    &#39;script/&#39;+config[&#39;script_coexp_module&#39;]</code></pre>
<ul>
<li>Steps 3-6,</li>
</ul>
<pre class="bash"><code>configfile: &quot;config.yaml&quot;

path=&#39;result/Nmodule.txt&#39;
with open(path) as f:
    lines = [x.rstrip() for x in f]
Nmodule=int(lines[0])

MODULE=list(range(1, Nmodule+1))
CHRS=list(range(1, config[&#39;Nchr&#39;]+1))
PERM=list(range(1, config[&#39;Nperm&#39;]+1))
fdr_thre_chr_module=config[&#39;fdr_level&#39;]

rule all:
  input:
    #expand(&#39;p/p.module{module}.chr{chr}.rds&#39;, module=MODULE, chr=CHRS)
    &#39;postanalysis/indep.signals.chr.module.perm&#39;+str(config[&#39;Nperm&#39;])+&#39;.txt&#39;

rule prep_bed:
  input:
    file_covariates=config[&#39;dir_expression&#39;]+config[&#39;file_covariates&#39;],
    file_gene_meta=&#39;result/&#39;+config[&#39;file_gene_meta&#39;],
    file_coexp_module=&#39;result/&#39;+config[&#39;file_coexp_module&#39;],
    file_ex=config[&#39;dir_expression&#39;]+config[&#39;file_ex&#39;]
  output:
    file_expression=expand(&#39;result/expression.module{module}.bed.gz&#39;, module=MODULE)
  params:
    data_type=&#39;obs&#39;
  script:
    &#39;script/&#39;+config[&#39;script_prep_bed&#39;]

rule z:
  input:
    expression_bed=&#39;result/expression.module{module}.bed.gz&#39;,
    file_covariates=config[&#39;dir_expression&#39;]+config[&#39;file_covariates&#39;]
  output:
    file_z=&#39;z/z.module{module}.chr{chr}.txt.gz&#39;
  params:
    plink_prefix_path=config[&#39;dir_geno&#39;]+config[&#39;geno_prefix&#39;]+&#39;{chr}&#39;+config[&#39;geno_suffix&#39;],
    prefix=&#39;module{module}.chr{chr}&#39;,
    dir_script=&#39;script/&#39;
  shell: &#39;bash &#39;+&#39;script/&#39;+config[&#39;script_z&#39;]+&#39; {params.plink_prefix_path} {input.expression_bed} {input.file_covariates} {params.prefix} {params.dir_script} {output.file_z}&#39;

rule p:
  input:
    file_ex_var_regressed=&#39;result/&#39;+config[&#39;file_ex_var_regressed&#39;],
    file_gene_meta=&#39;result/&#39;+config[&#39;file_gene_meta&#39;],
    file_coexp_module=&#39;result/&#39;+config[&#39;file_coexp_module&#39;],
    file_z=&#39;z/z.module{module}.chr{chr}.txt.gz&#39;
  output:
    file_p=&#39;p/p.module{module}.chr{chr}.rds&#39;
  params:
    dir_script=&#39;script/&#39;, chr=&#39;{chr}&#39;, module=&#39;{module}&#39;
  script:
    &#39;script/&#39;+config[&#39;script_p&#39;]

rule prep_bed_null:
  input:
    file_covariates=config[&#39;dir_expression&#39;]+config[&#39;file_covariates&#39;],
    file_gene_meta=&#39;result/&#39;+config[&#39;file_gene_meta&#39;],
    file_coexp_module=&#39;result/&#39;+config[&#39;file_coexp_module&#39;],
    file_ex=config[&#39;dir_expression&#39;]+config[&#39;file_ex&#39;]
  output:
    file_expression=temp(expand(&#39;result/expression.null.module{module}.perm{perm}.bed.gz&#39;, module=MODULE, allow_missing=True)),
    file_covariates_null=temp(&#39;result/covariates.null.perm{perm}.txt&#39;)
  params:
    data_type=&#39;null&#39;
  script:
    &#39;script/&#39;+config[&#39;script_prep_bed&#39;]


rule z_null:
  input:
    expression_bed=&#39;result/expression.null.module{module}.perm{perm}.bed.gz&#39;,
    file_covariates=&#39;result/covariates.null.perm{perm}.txt&#39;
  output:
    file_z=temp(&#39;z/z.null.module{module}.chr{chr}.perm{perm}.txt.gz&#39;)
  params:
    plink_prefix_path=config[&#39;dir_geno&#39;]+config[&#39;geno_prefix&#39;]+&#39;{chr}&#39;+config[&#39;geno_suffix&#39;],
    prefix=&#39;module{module}.chr{chr}.perm{perm}.null&#39;,
    dir_script=&#39;script/&#39;
  shell: &#39;bash script/&#39;+config[&#39;script_z&#39;]+&#39; {params.plink_prefix_path} {input.expression_bed} {input.file_covariates} {params.prefix} {params.dir_script} {output.file_z}&#39;


rule p_null:
  input:
    file_ex_var_regressed=&#39;result/&#39;+config[&#39;file_ex_var_regressed&#39;],
    file_gene_meta=&#39;result/&#39;+config[&#39;file_gene_meta&#39;],
    file_coexp_module=&#39;result/&#39;+config[&#39;file_coexp_module&#39;],
    file_z=&#39;z/z.null.module{module}.chr{chr}.perm{perm}.txt.gz&#39;
  output:
    file_p=&#39;p/p.null.module{module}.chr{chr}.perm{perm}.rds&#39;
  params:
    dir_script=&#39;script/&#39;, chr=&#39;{chr}&#39;, module=&#39;{module}&#39;
  script:
    &#39;script/&#39;+config[&#39;script_p&#39;]


rule FDR_chr_module:
  input:
    file_p=expand(&#39;p/p.module{module}.chr{chr}.rds&#39;,chr=CHRS,module=MODULE),
    file_p_null=expand(&#39;p/p.null.module{module}.chr{chr}.perm{perm}.rds&#39;,chr=CHRS,module=MODULE, allow_missing=True)
  output:
    file_q=&#39;FDR/q.chr.module.perm{perm}.rds&#39;
  script:
    &#39;script/&#39;+config[&#39;script_q&#39;]

rule average_perm_chr_module:
  input:
    file_q=expand(&#39;FDR/q.chr.module.perm{perm}.rds&#39;, perm=PERM)
  output:
    file_signals=&#39;FDR/signals.chr.module.perm&#39;+str(config[&#39;Nperm&#39;])+&#39;.txt&#39;
  params:
    fdr_thre=fdr_thre_chr_module
  script:
    &#39;script/&#39;+config[&#39;script_average_perm&#39;]


rule post_chr_module:
  input: sig=&#39;FDR/signals.chr.module.perm&#39;+str(config[&#39;Nperm&#39;])+&#39;.txt&#39;
  output: sig_uniq=&#39;postanalysis/LD.prun.in.chr.module.perm&#39;+str(config[&#39;Nperm&#39;])+&#39;.txt&#39;, sig_indp=&#39;postanalysis/indep.signals.chr.module.perm&#39;+str(config[&#39;Nperm&#39;])+&#39;.txt&#39;
  params: dir_geno=config[&#39;dir_geno&#39;], geno_prefix=config[&#39;geno_prefix&#39;], geno_suffix=config[&#39;geno_suffix&#39;]
  shell: &#39;bash script/&#39;+config[&#39;script_post&#39;]+&#39; {input.sig} {output.sig_uniq} {output.sig_indp} {params.dir_geno} {params.geno_prefix} {params.geno_suffix}&#39;</code></pre>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.2 (2021-11-01)
## Platform: x86_64-conda-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.3 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /scratch/midway2/liliw1/conda_env/rstudio-server/lib/libopenblasp-r0.3.18.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] forcats_0.5.1   stringr_1.4.0   dplyr_1.0.7     purrr_0.3.4    
## [5] readr_2.1.2     tidyr_1.2.0     tibble_3.1.7    ggplot2_3.3.6  
## [9] tidyverse_1.3.1
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_1.1.1 xfun_0.29        bslib_0.3.1      haven_2.4.3     
##  [5] colorspace_2.0-3 vctrs_0.4.1      generics_0.1.2   htmltools_0.5.2 
##  [9] yaml_2.2.2       utf8_1.2.2       rlang_1.0.3      jquerylib_0.1.4 
## [13] pillar_1.7.0     withr_2.5.0      glue_1.6.2       DBI_1.1.3       
## [17] dbplyr_2.1.1     modelr_0.1.8     readxl_1.3.1     lifecycle_1.0.1 
## [21] cellranger_1.1.0 munsell_0.5.0    gtable_0.3.0     rvest_1.0.2     
## [25] evaluate_0.14    knitr_1.37       tzdb_0.2.0       fastmap_1.1.0   
## [29] fansi_1.0.3      broom_0.7.12     Rcpp_1.0.8.3     backports_1.4.1 
## [33] scales_1.2.0     jsonlite_1.7.3   fs_1.5.2         hms_1.1.1       
## [37] digest_0.6.29    stringi_1.7.6    grid_4.1.2       cli_3.3.0       
## [41] tools_4.1.2      magrittr_2.0.3   sass_0.4.0       crayon_1.5.1    
## [45] pkgconfig_2.0.3  ellipsis_0.3.2   xml2_1.3.3       reprex_2.0.1    
## [49] lubridate_1.8.0  assertthat_0.2.1 rmarkdown_2.11   httr_1.4.2      
## [53] rstudioapi_0.13  R6_2.5.1         compiler_4.1.2</code></pre>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="unnumbered">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-battle2014" class="csl-entry">
Battle, Alexis, Sara Mostafavi, Xiaowei Zhu, James B. Potash, Myrna M. Weissman, Courtney McCormick, Christian D. Haudenschild, et al. 2014. <span>“Characterizing the Genetic Basis of Transcriptome Diversity Through RNA-Sequencing of 922 Individuals.”</span> <em>Genome Research</em> 24 (1): 14–24. <a href="https://doi.org/10.1101/gr.155192.113">https://doi.org/10.1101/gr.155192.113</a>.
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
