<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Run null simulations</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Trans-PCO</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="sim_null.html">Null simulations</a>
    </li>
    <li>
      <a href="sim_alt.html">Power comparison in simulations</a>
    </li>
    <li>
      <a href="rna_seq.html">Analyze a RNA-seq dataset</a>
    </li>
    <li>
      <a href="vis_distri_sig.html">Visualize trans signals</a>
    </li>
    <li>
      <a href="signal_cis.html">Cis genes of trans signals</a>
    </li>
    <li>
      <a href="coloc_gwas.html">Colocalization analysis of trans loci and GWAS</a>
    </li>
    <li>
      <a href="coloc_vis_prop.html">Colocalization proportion of trans-eQTLs</a>
    </li>
    <li>
      <a href="sldsc.html">Perform S-LDSC</a>
    </li>
    <li>
      <a href="vis_sldsc.html">Visualize the trait heritability enrichment</a>
    </li>
    <li>
      <a href="coloc_cis.html">Colocalization analysis of trans loci and cis genes</a>
    </li>
  </ul>
</li>
<li>
  <a href="ref.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liliw-w/Trans">
    <span class="fa fa-github"></span>
     
    Source
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Run null simulations</h1>

</div>


<div id="goal" class="section level2">
<h2>Goal</h2>
<p>Run null simulations to show trans-pco and the other two methods, i.e. minP and PC1, are well-calibrated.</p>
</div>
<div id="run-three-association-tests-on-simulated-null-z-scores" class="section level2">
<h2>Run three association tests on simulated null z-scores</h2>
<p>First, load required packages.</p>
<pre class="r"><code>rm(list = ls())
library(mvtnorm)
source(&quot;/home/liliw1/Trans/plot/theme_my_pub.R&quot;)</code></pre>
<p>Give input files and parameters. To make it simple (and fast), here I simulate only <span class="math inline">\(10^3\)</span> simulations (in the paper, <span class="math inline">\(10^7\)</span> tests were simulated).</p>
<pre class="r"><code>n_sim &lt;- 10^3</code></pre>
<pre class="r"><code># I/O &amp; paras -----
dir_pco &lt;- &#39;simulation/script_lambda0.1/&#39;
file_Sigma &lt;- &#39;/project2/xuanyao/llw/simulation_lambda0.1/new_Sigma/Sigma-new_DGN_module29_K101.rds&#39;</code></pre>
<p>Read files and source pco test scripts.</p>
<pre class="r"><code># read files -----
# source pco test
source(paste0(dir_pco, &quot;ModifiedPCOMerged.R&quot;))
source(paste0(dir_pco, &quot;liu.R&quot;))
source(paste0(dir_pco, &quot;liumod.R&quot;))
source(paste0(dir_pco, &quot;davies.R&quot;))
dyn.load(paste0(dir_pco, &quot;qfc.so&quot;))
source(paste0(dir_pco, &quot;ModifiedSigmaOEstimate.R&quot;))

Sigma &lt;- as.matrix(readRDS(file_Sigma))
K &lt;- dim(Sigma)[1]</code></pre>
<p>Simulate null z-scores using the given <span class="math inline">\(\Sigma\)</span> of a gene module from real dataset.</p>
<pre class="r"><code># simulate null z-scores -----
set.seed(123)

p_null_all  &lt;- list()
z_null &lt;- rmvnorm(n_sim, rep(0, K), Sigma)</code></pre>
<p>Run pco test and calculate p-values,</p>
<pre class="r"><code># prepare eigenvalues and eigenvectors as input for three methods -----
SigmaO &lt;- ModifiedSigmaOEstimate(Sigma)
eigen_res &lt;- eigen(Sigma)
eigen_lamb &lt;- eigen_res$values
eigen_vec &lt;- eigen_res$vectors

# PCO
p_null_all$&#39;p.null.PCO&#39; &lt;- ModifiedPCOMerged(
  Z.mat = z_null, Sigma = Sigma, SigmaO = SigmaO
) |&gt; as.numeric()</code></pre>
<pre><code>## PCMinP done. 
## PCFisher done. 
## PCLC done. 
## WI done. 
## Wald done. 
## VC done. 
## PCO done.</code></pre>
<pre class="r"><code>cat(&quot;PCO done. \n\n&quot;)</code></pre>
<pre><code>## PCO done.</code></pre>
<p>Run pc1 test and calculate p-values,</p>
<pre class="r"><code># PC1
PC1 &lt;- z_null %*% eigen_vec[, 1]
p_null_all$&#39;p.null.PC1&#39; &lt;- 2*pnorm(-abs(PC1/sqrt(eigen_lamb[1])))|&gt; as.numeric()

cat(&quot;PC1 done. \n\n&quot;)</code></pre>
<pre><code>## PC1 done.</code></pre>
<p>Run minp test and calculate p-values,</p>
<pre class="r"><code># univariate minp
p_null_all$&#39;p.null.minp&#39; &lt;- apply(z_null, 1, function(x) min(1-pchisq(x^2, 1))*length(x) )

cat(&quot;minp done. \n\n&quot;)</code></pre>
<pre><code>## minp done.</code></pre>
<p>Finally, take a look at the calculated p-values on null z’s.</p>
<pre class="r"><code>str(p_null_all)</code></pre>
<pre><code>## List of 3
##  $ p.null.PCO : num [1:1000] 0.867 0.561 0.287 0.312 0.724 ...
##  $ p.null.PC1 : num [1:1000] 0.349 0.633 0.771 0.162 0.679 ...
##  $ p.null.minp: num [1:1000] 1.313 0.871 1.747 0.457 1.101 ...</code></pre>
</div>
<div id="qq-plot-of-null-p-values-by-three-association-tests" class="section level2">
<h2>QQ-plot of null p-values by three association tests</h2>
<p>First, load required packages.</p>
<pre class="r"><code>rm(list = ls())
library(tidyverse)</code></pre>
<pre><code>## ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
<pre><code>## ✔ ggplot2 3.3.6     ✔ purrr   0.3.4
## ✔ tibble  3.1.7     ✔ dplyr   1.0.7
## ✔ tidyr   1.2.0     ✔ stringr 1.4.0
## ✔ readr   2.1.2     ✔ forcats 0.5.1</code></pre>
<pre><code>## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<pre class="r"><code>source(&quot;/home/liliw1/Trans/plot/theme_my_pub.R&quot;)</code></pre>
<p>Files of simulated null p-values,</p>
<pre class="r"><code># I/O &amp; paras -----
file_dat_null &lt;- &quot;/project2/xuanyao/llw/simulation_lambda0.1/new_Sigma/simulation.null.lambda0.1.K101.rds&quot;
ci_level &lt;- 0.95</code></pre>
<pre class="r"><code># read files -----
p_null_all &lt;- readRDS(file_dat_null)



# organize data -----
## reset obs p-values that are 0 to a fixed value, here I use the min non-zero p/10 -----
input &lt;- as_tibble(p_null_all)
input[input == 0] &lt;- min(input[input != 0])/10

## number of samples -----
n &lt;- nrow(input)

## expected null p -----
expected &lt;- seq(1, n) / (n+1)
lexp &lt;- -log10(expected)

## order statistic of null p -----
ci_l &lt;- -log10( qbeta(p = (1 - ci_level) / 2, shape1 = 1:n, shape2 = n:1) )
ci_r &lt;- -log10( qbeta(p = (1 + ci_level) / 2, shape1 = 1:n, shape2 = n:1) )

## obs -----
observed &lt;- apply(input, 2, sort) %&gt;% as.data.frame()
lobs &lt;- -log10(observed)


## take only a subset of null p&#39;s, to save image space -----
ind_sub &lt;- c(
  1:sum(lexp &gt; 4),
  seq(from = sum(lexp &gt; 4), to = sum(lexp &gt; 2), length.out = 2000) %&gt;% ceiling(),
  seq(from = sum(lexp &gt; 2), to = n, length.out = 2000) %&gt;% ceiling()
)


# data for plt -----
df_plt &lt;- cbind(data.frame(x = lexp, ci_l = ci_l, ci_r = ci_r), lobs) %&gt;%
  slice(ind_sub) %&gt;%
  pivot_longer(-c(x, ci_l, ci_r), names_to = &quot;Type&quot;, values_to = &quot;y&quot;)

## set group order in plt -----
group_order &lt;- c(&quot;p.null.PCO&quot;, &quot;p.null.PC1&quot;, &quot;p.null.minp&quot;)
group_label &lt;- c(&quot;Trans-PCO&quot;, &quot;PC1&quot;, &quot;MinP&quot;)

group_order &lt;- if(is.null(group_order)) unique(df_plt$Type) else group_order
group_label &lt;- if(is.null(group_label)) group_order else group_label
df_plt$Type &lt;- factor(df_plt$Type, levels = group_order, labels = group_label)


# QQ-plot of null p-values by three association tests -----
base_plt &lt;- ggplot(df_plt, aes(x = x, y = y, group = Type)) +
  geom_ribbon(aes(ymin = ci_l, ymax = ci_r), fill = &quot;#e5e5e5&quot;, color = &quot;#e5e5e5&quot;) +
  geom_abline(slope = 1, intercept = 0, color = &quot;#595959&quot;, size = 0.7) +
  geom_point(aes(color = Type), size = 0.5) +
  labs(x = bquote(Expected -log[10]~italic((P))),
       y = bquote(Observed -log[10]~italic((P))),
       color = NULL)
base_plt +
  scale_color_manual(
    values = c(&quot;#85192d&quot;, &quot;#0028a1&quot;, &quot;#e89c31&quot;),
    guide = guide_legend(override.aes = list(size = 2))
  ) +
  theme_my_pub() +
  theme(
    panel.grid.major.x = element_line(linetype = &quot;dotted&quot;),
    panel.grid.major.y = element_line(linetype = &quot;dotted&quot;),
    
    legend.background = element_blank(),
    legend.position = &quot;right&quot;,
    
    axis.title = element_text(size = 14), 
    axis.text = element_text(colour = &quot;black&quot;, size = 12)
  )</code></pre>
<p><img src="sim_null_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.2 (2021-11-01)
## Platform: x86_64-conda-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.3 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /scratch/midway2/liliw1/conda_env/rstudio-server/lib/libopenblasp-r0.3.18.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] forcats_0.5.1   stringr_1.4.0   dplyr_1.0.7     purrr_0.3.4    
##  [5] readr_2.1.2     tidyr_1.2.0     tibble_3.1.7    ggplot2_3.3.6  
##  [9] tidyverse_1.3.1 mvtnorm_1.1-3  
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.8.3     lubridate_1.8.0  assertthat_0.2.1 digest_0.6.29   
##  [5] utf8_1.2.2       R6_2.5.1         cellranger_1.1.0 backports_1.4.1 
##  [9] reprex_2.0.1     evaluate_0.14    httr_1.4.2       highr_0.9       
## [13] pillar_1.7.0     rlang_1.0.3      readxl_1.3.1     rstudioapi_0.13 
## [17] jquerylib_0.1.4  rmarkdown_2.11   labeling_0.4.2   munsell_0.5.0   
## [21] broom_0.7.12     compiler_4.1.2   modelr_0.1.8     xfun_0.29       
## [25] pkgconfig_2.0.3  htmltools_0.5.2  tidyselect_1.1.1 fansi_1.0.3     
## [29] crayon_1.5.1     tzdb_0.2.0       dbplyr_2.1.1     withr_2.5.0     
## [33] grid_4.1.2       jsonlite_1.7.3   gtable_0.3.0     lifecycle_1.0.1 
## [37] DBI_1.1.3        magrittr_2.0.3   scales_1.2.0     cli_3.3.0       
## [41] stringi_1.7.6    farver_2.1.1     fs_1.5.2         xml2_1.3.3      
## [45] bslib_0.3.1      ellipsis_0.3.2   generics_0.1.2   vctrs_0.4.1     
## [49] tools_4.1.2      glue_1.6.2       hms_1.1.1        fastmap_1.1.0   
## [53] yaml_2.2.2       colorspace_2.0-3 rvest_1.0.2      knitr_1.37      
## [57] haven_2.4.3      sass_0.4.0</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
