<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Run simulations under alternatives</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Trans-PCO</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="sim_null.html">Null simulations</a>
    </li>
    <li>
      <a href="sim_alt.html">Power comparison in simulations</a>
    </li>
    <li>
      <a href="rna_seq.html">Analyze a RNA-seq dataset</a>
    </li>
    <li>
      <a href="vis_distri_sig.html">Visualize trans signals</a>
    </li>
    <li>
      <a href="signal_cis.html">Cis genes of trans signals</a>
    </li>
    <li>
      <a href="coloc_gwas.html">Colocalization analysis of trans loci and GWAS</a>
    </li>
    <li>
      <a href="coloc_vis_prop.html">Colocalization proportion of trans-eQTLs</a>
    </li>
    <li>
      <a href="sldsc.html">Perform S-LDSC</a>
    </li>
    <li>
      <a href="vis_sldsc.html">Visualize the trait heritability enrichment</a>
    </li>
    <li>
      <a href="coloc_cis.html">Colocalization analysis of trans loci and cis genes</a>
    </li>
  </ul>
</li>
<li>
  <a href="ref.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liliw-w/Trans">
    <span class="fa fa-github"></span>
     
    Source
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Run simulations under alternatives</h1>

</div>


<div id="goal" class="section level2">
<h2>Goal</h2>
<p>This page is to visualize the identified trans-eQTLs in the RNA-seq dataset.</p>
</div>
<div id="look-at-the-co-expression-modules-and-module-sizes" class="section level2">
<h2>Look at the co-expression modules and module sizes</h2>
<pre class="r"><code>file_coexp_module &lt;- &quot;/project2/xuanyao/llw/DGN_no_filter_on_mappability/result/coexp.module.rds&quot;
coexp_module &lt;- readRDS(file_coexp_module)$moduleLabels


# modules and their sizes
df_module &lt;- tibble(&quot;gene&quot; = names(coexp_module),
                       &quot;module&quot; = coexp_module)
df_module_size &lt;- df_module %&gt;% group_by(module) %&gt;% summarise(module_size = n())</code></pre>
<p>There are 166 co-expression modules in total, consisting of 12,132 genes.</p>
<pre class="r"><code>cat(
  &#39;There are&#39;, max(df_module$module),
  &#39;co-expression modules in total, consisting of&#39;,
  nrow(df_module), &#39;genes. \n\n&#39;
)</code></pre>
<pre><code>## There are 166 co-expression modules in total, consisting of 12132 genes.</code></pre>
<p>The module sizes are distributed as below.</p>
<pre class="r"><code># histogram of number of modules v.s. module size
base_plt &lt;- ggplot(df_module_size) +
  geom_histogram(aes(x = module_size, fill = after_stat(count)),
                 binwidth = 10,
                 color = &quot;white&quot;) +
  scale_fill_gradientn(colors = RColorBrewer::brewer.pal(8, &quot;Blues&quot;)[8:4],
                       na.value = &quot;red&quot;) +
  labs(x = &quot;Module Size&quot;, y = &quot;Number of Modules&quot;)

base_plt +
  theme_classic() +
  theme(panel.grid.major.y = element_line(linetype = &quot;dashed&quot;),
        panel.background = element_rect(fill = &quot;white&quot;),
        
        legend.position = &quot;none&quot;,
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 10, face = &quot;bold&quot;),
        legend.background = element_rect(color = &quot;black&quot;, linetype = &quot;dashed&quot;),
        legend.key.size= unit(0.5, &quot;cm&quot;),
        
        axis.line = element_line(colour=&quot;black&quot;, size = 0.7),
        axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        
        plot.margin=unit(c(10,5,5,5),&quot;mm&quot;),
        axis.text.x = element_text(colour=&quot;black&quot;, hjust=1, vjust = 1, size = 12),
        axis.text.y = element_text(colour = &quot;black&quot;, size = 12),
        axis.title.x = element_text(vjust = -0.2, size=14),
        axis.title.y = element_text(angle=90,vjust =2, size=14)
  )</code></pre>
<p><img src="vis_distri_sig_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
</div>
<div id="trans-signals-and-their-corresponding-trans-target-modules" class="section level2">
<h2>Trans signals and their corresponding trans target modules</h2>
<p>Figure below: Signals of each module on chromosomes. Y-axis is chromosome position. X-axis shows module. Each point is a signal. Point size means -logP.</p>
<pre class="r"><code># I/O &amp; paras -----
file_chr_pos &lt;- &#39;/scratch/midway2/liliw1/sig_module_chr/chromosome_location.rds&#39;
file_qtl &lt;- &#39;/project2/xuanyao/llw/DGN_no_filter_on_mappability/FDR/signals.chr.module.perm10.fdr10.txt&#39;


# read files -----
chr_pos &lt;- readRDS(file_chr_pos)
qtl &lt;- fread(file_qtl, header = FALSE, col.names = c(&quot;signal&quot;, &quot;p&quot;, &quot;q&quot;))


# organize data -----
qtl &lt;- qtl %&gt;%
  separate(signal, c(&quot;module&quot;, &quot;chr&quot;, &quot;pos&quot;), &quot;:&quot;, remove = FALSE, convert = TRUE) %&gt;%
  unite(&quot;SNP_ID&quot;, chr, pos, sep = &quot;:&quot;, remove = FALSE) %&gt;%
  separate(module, c(NA, &quot;module&quot;), sep = &#39;module&#39;, convert = TRUE)


# plot -logp for (cht pos, module) -----
plt_dat &lt;- qtl %&gt;%
  left_join(chr_pos, by = c(&quot;chr&quot; = &quot;CHR&quot;)) %&gt;%
  mutate(def_pos = pos + tot)

base_plt &lt;- ggplot(plt_dat, aes(y = def_pos, x = factor(module))) +
  geom_rect(
    aes(ymin = tot, ymax = xmax,
        xmin = -Inf, xmax = Inf,
        fill = factor(chr))
  ) +
  geom_vline(aes(xintercept = factor(module)),
             linetype = &quot;dotted&quot;, color = &quot;#e5e5e5&quot;) +
  geom_point(aes(size = -log10(p), color = factor(chr)),
             alpha = 0.5, shape = 1) +
  labs(x = &quot;Module&quot;, y = &quot;Chromosome&quot;, size = quote(-Log[10](P)))

base_plt +
  scale_y_continuous(
    limits = c(0, max(chr_pos$center)*2 - max(chr_pos$tot)),
    label = chr_pos$CHR,
    breaks = chr_pos$center,
    expand = c(0, 0)
  ) +
  scale_size(guide = guide_legend(override.aes = list(alpha = 1)), range = c(0.5, 3)) +
  scale_color_manual(values = rep(c(&quot;#843232&quot;, &quot;#000099&quot;), 22), guide = &quot;none&quot;) +
  scale_fill_manual(values = rep(c(&quot;#e5e5e5&quot;, &quot;#ffffff&quot;), 22), guide = &quot;none&quot;) +
  theme_my_pub() +
  theme(
    axis.text.x = element_text(angle = 90, size = 8)
  )</code></pre>
<p><img src="vis_distri_sig_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
</div>
<div id="distribution-of-trans-signals-on-their-corresponding-target-modules" class="section level2">
<h2>Distribution of trans signals on their corresponding target modules</h2>
<pre class="r"><code>ggplot(qtl) +
  geom_bar(aes(x = as.factor(module))) +
  labs(x = &quot;Module&quot;, y = &quot;Number of trans signals&quot;) +
  theme_classic() +
  theme(
    axis.text.x = element_text(angle = 90, size = 8)
  )</code></pre>
<p><img src="vis_distri_sig_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
</div>
<div id="distribution-of-trans-signals-on-chromosomes" class="section level2">
<h2>Distribution of trans signals on chromosomes</h2>
<pre class="r"><code>ggplot(qtl) +
  geom_bar(aes(x = as.factor(chr))) +
  labs(x = &quot;Chromosome&quot;, y = &quot;Number of trans signals&quot;) +
  theme_classic() +
  theme(
    axis.text.x = element_text(angle = 90, size = 8)
  )</code></pre>
<p><img src="vis_distri_sig_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.2 (2021-11-01)
## Platform: x86_64-conda-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.3 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /scratch/midway2/liliw1/conda_env/rstudio-server/lib/libopenblasp-r0.3.18.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] forcats_0.5.1     stringr_1.4.0     dplyr_1.0.7       purrr_0.3.4      
##  [5] readr_2.1.2       tidyr_1.2.0       tibble_3.1.7      ggplot2_3.3.6    
##  [9] tidyverse_1.3.1   data.table_1.14.2
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_1.1.1   xfun_0.29          bslib_0.3.1        haven_2.4.3       
##  [5] colorspace_2.0-3   vctrs_0.4.1        generics_0.1.2     htmltools_0.5.2   
##  [9] yaml_2.2.2         utf8_1.2.2         rlang_1.0.3        jquerylib_0.1.4   
## [13] pillar_1.7.0       withr_2.5.0        glue_1.6.2         DBI_1.1.3         
## [17] RColorBrewer_1.1-3 dbplyr_2.1.1       modelr_0.1.8       readxl_1.3.1      
## [21] lifecycle_1.0.1    cellranger_1.1.0   munsell_0.5.0      gtable_0.3.0      
## [25] rvest_1.0.2        evaluate_0.14      labeling_0.4.2     knitr_1.37        
## [29] tzdb_0.2.0         fastmap_1.1.0      fansi_1.0.3        highr_0.9         
## [33] Rcpp_1.0.8.3       broom_0.7.12       backports_1.4.1    scales_1.2.0      
## [37] jsonlite_1.7.3     farver_2.1.1       fs_1.5.2           hms_1.1.1         
## [41] digest_0.6.29      stringi_1.7.6      grid_4.1.2         cli_3.3.0         
## [45] tools_4.1.2        magrittr_2.0.3     sass_0.4.0         crayon_1.5.1      
## [49] pkgconfig_2.0.3    ellipsis_0.3.2     xml2_1.3.3         reprex_2.0.1      
## [53] lubridate_1.8.0    assertthat_0.2.1   rmarkdown_2.11     httr_1.4.2        
## [57] rstudioapi_0.13    R6_2.5.1           compiler_4.1.2</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
